### PR TITLE
Support FlashMLA sparse prefill with DSA context parallel

### PR DESCRIPTION
**Pull request by andrewdoro**
## Summary
- extend DeepSeek V4 topk v2 to optionally emit raw positional indices
- route FlashMLA sparse prefill through the CP-compatible v2 topk path
- add CP round-robin row/position handling for sparse prefill
- fall back to dense prefill for padded CP query rows or multi-request CP chunks

## Validation
- Runpod B200: topk v2 raw indices match v1 for topk 512/1024 on short and 33k sequence cases
- Runpod B200: sparse prefill combine uses explicit CP positions and emits expected combined indices/lens
- Runpod/local: sparse prefill shape gate regression test passes
- Modal: prior iter9s boot + long prefill smoke passed without Triton device-side assert; full benchmarks still pending

Draft until the 8x B200 Modal benchmark/accuracy pass is complete.

---
### CI States

Latest PR Test (Base): :x: [Run #26363685914](https://github.com/sgl-project/sglang/actions/runs/26363685914)
Latest PR Test (Extra): :no_entry_sign: [Run #26363889259](https://github.com/sgl-project/sglang/actions/runs/26363889259)

### CODE REVIEW

**Comment by andrewdoro** [comment]
Validation update from the CP port branch:

- Local: `python3 -m py_compile python/sglang/srt/layers/attention/deepseek_v4_backend.py` passed.
- Local: `PYTHONPATH=/private/tmp/sglang-pr25418-cp-port/python python3 -m pytest -q test/srt/layers/attention/dsv4/test_sparse_prefill_shape_gate.py` passed: 3 tests.
- Runpod H100: sparse-prefill combine registered test passed; topk-v2 raw-indices test is blocked on the Runpod CUDA 12.8 image failing to compile upstream `sgl_kernel/deepseek_v4/topk/ptx.cuh` mbarrier/cp_async_bulk intrinsics.
- Modal 8x B200 + DSA CP round-robin: booted cleanly with the patch, no Triton device-side assert. A 50k cold request completed: 52,029 prompt tokens, first token 2.382s. Instrumentation confirmed sparse prefill enabled for `compress_ratio=4` and `compress_ratio=128` across all 8 CP ranks.

Latest pushed commit: `29a59bd82`.

File: python/sglang/srt/layers/attention/dsv4/metadata.py [hunk 1]
```diff
@@ -48,6 +48,7 @@
     c4_sparse: means "compressed by 4" but only attend to top-512 tokens.
                all related length will be clipped to 512.
 """
+_LARGE_INDEXER_QUERY_THRESHOLD = 11673
```
[the remark this answers is not in the document]
Comment (andrewdoro) [reply python/sglang/srt/layers/attention/dsv4/metadata.py:52]: Fixed in 8e2f6756. The threshold now has a one-line comment tying it to the FlashMLA decode scheduler shared-memory cap: `4 * (b * 5 + 1) <= 228 KiB`.

File: python/sglang/srt/layers/attention/dsv4/sparse_prefill_utils.py [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+    num_reqs = seq_lens.shape[0]
+    device = seq_lens.device
+
+    swa_gather_lens = torch.minimum(seq_lens, extend_seq_lens + (swa_window - 1)).to(
+        torch.int32
+    )
+    swa_first_pos = (seq_lens - swa_gather_lens).to(torch.int32)
+    swa_offsets = torch.zeros(num_reqs + 1, dtype=torch.int32, device=device)
+    swa_offsets[1:] = torch.cumsum(swa_gather_lens, dim=0).to(torch.int32)
+    total_swa = int(swa_offsets[-1].item())  # one CPU sync per chunk
```
[the remark this answers is not in the document]
Comment (andrewdoro) [reply python/sglang/srt/layers/attention/dsv4/sparse_prefill_utils.py:214]: Keeping this as-is for now. The sparse workspace needs a host-side `total_swa` to size temporary tensors before launching FlashMLA sparse prefill, and this path runs once per sparse prefill chunk. I called it out in the PR validation notes as something to revisit if profiling shows the sync eating the speedup.

File: python/sglang/srt/layers/attention/dsv4/dequant_k_cache.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        scale_pow2 = tl.exp2((scale_u8 - 127).to(tl.float32))
+
+        out_off = out_row_base + tile_id * TILE_SIZE + nope_offs
+        tl.store(
+            output_ptr + out_off,
+            (fp8_vals * scale_pow2).to(output_ptr.dtype.element_ty),
+        )
+
+    rope_offs = tl.arange(0, DIM_ROPE)
+    bf16_off = (token_data_base + DIM_NOPE) // 2 + rope_offs
```
[the remark this answers is not in the document]
Comment (andrewdoro) [reply python/sglang/srt/layers/attention/dsv4/dequant_k_cache.py]: Fixed in 8e2f6756. I added explicit even-alignment assertions and split the byte offset from the bf16 element offset so future constant changes fail loudly.

**Comment by andrewdoro** [comment]
Additional Modal validation after the patch-stack refresh:

- Modal 4x B200, `tp=4`, DSA prefill CP round-robin, EAGLE 3/1/4, mixed-chunk, `SGLANG_OPT_FLASHMLA_SPARSE_PREFILL=1` booted cleanly.
- Static patch applied cleanly after the Anara Modal stack (`PR#26186`, `PR#26201`).
- Fixed cold request returned 200: 43,917 prompt tokens, 2.114s wall, ~20.8k prompt tok/s wall.
- Logs confirmed `DSV4 sparse prefill enabled` for `compress_ratio=0`, `4`, and `128` across CP ranks 0-3.
- No Triton device-side assert / SIGQUIT. Test Modal app was stopped after the smoke.

Full 8x B200 200k/500k benchmark + accuracy pass is still pending, so keeping this draft.
